opte/src/engine/ip6.rs

```diff
@@ -209,49 +210,116 @@ impl Header for Ipv6Hdr {
+        let dst = Ipv6Addr::from(base_header.dst_addr().0);
+
+        // Parse any extension headers.
+        //
+        // At this point, we don't need any information out of the headers other
+        // than their length (to determine the boundary with the ULP). We'll
+        // verify that the headers are supported, but otherwise maintain only a
+        // byte array with their contents.
+        let mut extension_headers = vec![];
+        let mut next_header = next_hdr;
+        while !is_ulp_protocol(next_header) {
+            match next_header {
+                IpProtocol::HopByHop => {
+                    let buf = rdr.slice(rdr.seg_left())?;
+                    let header = Ipv6HopByHopHeader::new_checked(buf)?;
+                    let n_bytes = 8 * (usize::from(header.header_len()) + 1);
+                    extension_headers.extend_from_slice(&buf[..n_bytes]);
+                    next_header = header.next_header();
+
+                    // Put back any bytes in the segment not needed for this header.
+                    rdr.seek_back(buf.len() - n_bytes)?;
+                }
+                IpProtocol::Ipv6Route => {
+                    let buf = rdr.slice(rdr.seg_left())?;
+                    let header = Ipv6RoutingHeader::new_checked(buf)?;
+                    let n_bytes = 8 * (usize::from(header.header_len()) + 1);
+                    extension_headers.extend_from_slice(&buf[..n_bytes]);
+                    next_header = header.next_header();
+                    rdr.seek_back(buf.len() - n_bytes)?;
+                }
+                IpProtocol::Ipv6Frag => {
+                    // This header's length is fixed
+                    let buf = rdr.slice(8)?;
+                    let header = Ipv6FragmentHeader::new_checked(buf)?;
+                    extension_headers.extend_from_slice(buf);
+                    next_header = header.next_header();
+                }
+                IpProtocol::Unknown(x) if x == DDM_HEADER_ID => {
+                    // The DDM header packet begins with next_header and the
+                    // length, which describes the entire header excluding
+                    // next_header.
+                    const FIXED_LEN: usize = 2;
+                    let fixed_buf = rdr.slice(FIXED_LEN)?;
+                    next_header = IpProtocol::from(fixed_buf[0]);
+                    let total_length = usize::from(fixed_buf[1]) + 1;
+                    let remainder = rdr.slice(total_length - FIXED_LEN)?;
+                    extension_headers.extend_from_slice(fixed_buf);
+                    extension_headers.extend_from_slice(remainder);
+                }
+                x => {
+                    return Err(Ipv6HdrError::UnexpectedNextHeader {
+                        next_header: x.into(),
+                    });
+                }
+            }
+        }
+        // Safety: The protocol is the last value of next header, and since
```

Comment:
For the comments around `unwrap()` I've been using an `Unwrap:` tag (that's not totally true, there's tons of unwrap that has no annotation). Now I'm wondering if instead it should be `Panic:`, with the idea being that we document why we do not expect a panic -- which typically would involve the fact that we have information that we can't give to the compiler. Thoughts?

Reply:
I think `Panic:` is a good choice. I've seen `Safety` in a lot of places, but it's usually explain the explicit use of an `unsafe` block. I'll switch it.